### PR TITLE
.Net Processes: fixing KernelProcessStateMetadata usage

### LINKED ISSUE #9599: Improvement:  Improving usability of KernelProcessStepStateMetadata
Improving how  KernelProcessStepStateMetadata is structured to make it easier to adopt for future special steps/processes

### PR DESCRIPTION
**Pull request by esttenorio**
### Description

- fixing definition of `KernelProcessStateMetadata` to be easier to inherit/manage
- fixing missing plumbing of version in KernelState
- Updating Build interface in `ProcessStepBuilder` and `ProcessBuilder`

Fixes #9599 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Experimental/Process.Abstractions/Models/KernelProcessStepStateMetadata.cs [hunk 1]
```diff
@@ -8,6 +8,9 @@ namespace Microsoft.SemanticKernel.Process.Models;
 /// <summary>
 /// Step state used for State Persistence serialization
 /// </summary>
+[JsonPolymorphic(TypeDiscriminatorPropertyName = "$type", UnknownDerivedTypeHandling = JsonUnknownDerivedTypeHandling.FallBackToNearestAncestor)]
+[JsonDerivedType(typeof(KernelProcessStepStateMetadata), typeDiscriminator: nameof(KernelProcessStepStateMetadata))]
```
Comment (crickman) [inline dotnet/src/Experimental/Process.Abstractions/Models/KernelProcessStepStateMetadata.cs:12]: (Wondering) Considering the scope of this hierarachy (closed system / limited number of types), I might consider a more concise type discriminator that isn't bound to the type name: `Process`, `Step`, `Map` ?  (You are already aliasing property names so they aren't directly bound to model naming.)

Ultimately, either approach is just fine.

File: dotnet/src/Experimental/Process.Core/ProcessBuilder.cs [hunk 2]
```diff
@@ -96,31 +96,15 @@ internal override Dictionary<string, KernelFunctionMetadata> GetFunctionMetadata
     /// </summary>
     /// <param name="stateMetadata">State to apply to the step on the build process</param>
     /// <returns></returns>
-    internal override KernelProcessStepInfo BuildStep(KernelProcessStepStateMetadata<object>? stateMetadata)
+    internal override KernelProcessStepInfo BuildStep(KernelProcessStepStateMetadata? stateMetadata = null)
     {
         // The step is a, process so we can return the step info directly.
         if (stateMetadata is KernelProcessStateMetadata processState)
```
Comment (crickman) [inline dotnet/src/Experimental/Process.Core/ProcessBuilder.cs:102]: Nit / Not Blocking - I'm wondering if there's ever a cae where `stateMetadata` might not be `KernelProcessStateMetadata`

**Review by crickman — APPROVED** [review]

File: dotnet/src/Experimental/Process.Core/ProcessBuilder.cs (hunk 2 above)
Comment (esttenorio) [reply dotnet/src/Experimental/Process.Core/ProcessBuilder.cs:102]: I guess now the base class will always be `KernelProcessStepStateMetadata` regardless if a new statemetadata gets added or not.

File: dotnet/src/Experimental/Process.Abstractions/Models/KernelProcessStepStateMetadata.cs (hunk 1 above)
Comment (esttenorio) [reply dotnet/src/Experimental/Process.Abstractions/Models/KernelProcessStepStateMetadata.cs]: added changes in last commit to change the type discriminator

**Review by crickman — APPROVED** [review]

**Pull request merged**